driver/transform-dsl.md

```diff
@@ -0,0 +1,549 @@
+
+```json
+{
+    "type": "Ident",
+    "name": null,
+    "offset": 5
+}
+```
+
+As you can see, an operation can be used to describe a tree shape and
+reconstruct it back, assuming that all placeholder variables used in
+an operation are defined in node state.
+
+### Mappings
+
+Since mapping defines two operations (source and target), we can change
+our code to use one operation for checking the tree and building
+the state and a different operation for constructing it:
+
+```golang
+A := Part("other", Obj{
+    "type": String("Ident"),
+    "name": Var("x"),
+})
+
+B := Part("other", Obj{
+    "type": String("go:Ident"),
+    "token": Var("x"),
+})
+
+s := NewState()
+
+// verify an AST with operation A, populate state
+if ok, _ := A.Check(s, ast); ok {
+    // construct a new AST using the state and operation B
+    newAST, _ := B.Construct(s, nil)
+}
+```
+
+As long as all variable defined in A exactly match all variables defined
+in B, `Check`-`Construct` routine will work the same way as described
+in example section above.
+
+Since state only stores placeholder values, while operation defines the
+shape of the tree, any complex shape transformation can be defined:
+
+```golang
+Obj{
+    "type": String("Binary"),
+    "op": Var("op"),
+    "vals": Arr( Var("left"), Var("right") ),
+}
+
+// state: {"op": ..., "left": ..., "right": ...}
+
+Obj{
+    "type": String("Binary"),
+    "op": Obj{
+        "type": String("Operation"),
+        "token": Var("op")
+    },
+    "left": Var("left"),
+    "right": Var("right"),
+}
+```
+
+This example will transform this AST (matches the first transform):
+
+```json
+{
+    "type": "Binary",
+    "op": "+",
+    "vals": [
+        {"type":"Ident", "name": "v"},
+        5
+    ]
+}
+```
+
+Into a new AST (defined by the second transform):
+
+```json
+{
+    "type": "Binary",
+    "op": {
+        "type": "Operation",
+        "token": "+"
+    },
+    "left": {"type":"Ident", "name": "v"},
+    "right": 5
+}
+```
+
+
+## Restrictions
+
+DSL also enforces few restrictions on transformations and operations.
+
+### Completeness
+
+All operations should either:
+
+* Reject the tree if it does not match one of constraints.
+
+* Describe the whole tree, and save unknown/placeholder nodes to the state.
+
+This rule ensures that if an operation matches the tree, it would be able
+to replicate it from a state associated with the node, since all fields
+are covered by it.
+
+This property allows to test each transformation by executing
+`Check`-`Construct` routine, or detect if a mapping defined by a driver
+developer ignores certain AST fields that are seen in the wild
+(will result in an error).
+
+### Reversibility
+
+Since all operations are required to implement both `Check` and `Construct`
+they can be used interchangeably as either source and/or target transforms.
+
+In simple case it means that transform A can produce the state S for AST
```

Comment:
I would rephrase as:

"For a simple case, this means, that given a node T over which we apply a transform A, we can produce a state S (with `Check`) and then reconstruct the original node T (with `Construct`) by only using the state S."